### ISSUE TITLE
Remove `pending` in favor of `status` in data fetching composables

### ISSUE THREAD

**Issue by DamianGlowala** [nuxt/nuxt#25225]
### Describe the feature

Currently, data fetching composables expose pending information in two ways: `pending` and `status` (`status.value === 'pending'`). It would be great to consider removing `pending` for consistency, which currently has a duplicated source of truth. As this would be a breaking change, it's probably best to address it in the next version of Nuxt.

([Update in this comment](https://github.com/nuxt/nuxt/issues/25225#issuecomment-2182690498))

### Additional information

- [X] Would you be willing to help implement this feature?
- [ ] Could this feature be implemented as a module?

### Final checks

- [X] Read the [contribution guide](https://nuxt.com/docs/community/contribution).
- [X] Check existing [discussions](https://github.com/nuxt/nuxt/discussions) and [issues](https://github.com/nuxt/nuxt/issues).

**Comment by drowhannn** [comment]
@danielroe @DamianGlowala Can i work on this?

**Comment by danielroe** [comment]
You're welcome to open a PR but we probably won't merge it until we're ready to migrate to Nuxt 4.

You might also consider initial work on writing a codemod to migrate people's code to the new format.

**Comment by drowhannn** [comment]
@danielroe I am a little confused on how to replace `pending.value` in docs and tests. Should i replace it with `status.value = "pending"` or `status.value = "idle" || status.value = "pending" || !data.value`

**Comment by ctigrisht** [comment]
why not just use pending as a computed value of `status.value === "pending"` so you don't break compatibility?

**Comment by fen89** [comment]
@danielroe  Wouldn't it be better to keep `pending` as a computed, as @ctigrisht said - this is a convenient wrapper for useFetch; if this doesn't exist anymore then we need to add the good old boilerplate stuff again and again everywhere.

Imho the following comment from @drowhannn already shows that it is a bad idea to remove the flag, because everybody will be confused as well with this change.

> I am a little confused on how to replace pending.value in docs and tests. Should i replace it with status.value === "pending" or status.value === "idle" || status.value === "pending" || !data.value

To sum up: Why removing a convenient well established flag?

**Comment by yixiaco** [comment]
I want to be able to use status.pending, instead of status === 'pending'?

**Comment by manniL** [comment]
> I am a little confused on how to replace `pending.value` in docs and tests. Should i replace it with `status.value === "pending"` or `status.value === "idle" || status.value === "pending" || !data.value`

@drowhannn 

If you want the same behavior as before, use `status.value === "idle" || status.value === "pending" || !data.value`. If you want the "accurate" behavior of "when the request was triggered and is pending", use `status.value === "pending"` 



> why not just use pending as a computed value of `status.value === "pending"` so you don't break compatibility?

> To sum up: Why removing a convenient well established flag?

@fen89 @ctigrisht Would be an option but will break code without people realizing.

**Comment by bitbytebit1** [comment]
I'm strongly against removing the pending variable. It's incredibly convenient to use and a enables a very common pattern. It works beautifully with Suspense and v-if, is very concise to write and read.

**Comment by fen89** [comment]
> > I am a little confused on how to replace `pending.value` in docs and tests. Should i replace it with `status.value === "pending"` or `status.value === "idle" || status.value === "pending" || !data.value`
> 
> @drowhannn
> 
> If you want the same behavior as before, use `status.value === "idle" || status.value === "pending" || !data.value`. If you want the "accurate" behavior of "when the request was triggered and is pending", use `status.value === "pending"`
> 
> > why not just use pending as a computed value of `status.value === "pending"` so you don't break compatibility?
> 
> > To sum up: Why removing a convenient well established flag?
> 
> @fen89 @ctigrisht Would be an option but will break code without people realizing.


@manniL Isn't the whole issue about introducing a breaking change? 

> As this would be a breaking change, it's probably best to address it in the next version of Nuxt.

Before removing such flags there could be other options which imho should be at least considered. If we assume as the author said that this will introduce a breaking change why not checking if the code underneath can be looked at. 

Let's use Tanstack Query as a well thought api example, so that everybody can see that both worlds are possible.

Directly from the [Website](https://tanstack.com/query/latest/docs/framework/vue/guides/queries):

The result object contains a few very important states you'll need to be aware of to be productive. A query can only be in one of the following states at any given moment:

| variable | status equivalent | description |
| - | - | - |
| `isPending ` | `status === 'pending' ` | The query has no data yet |
| `isError ` |  `status === 'error'` |  The query encountered an error |
| `isSuccess` | `status === 'success'` |  The query was successful and data is available |

Beyond those primary states, more information is available depending on the state of the query:

| variable | description |
| - | - |
| `error` |  If the query is in an isError state, the error is available via the error property. |
| `data` |  If the query is in an isSuccess state, the data is available via the data property. |
| `isFetching` | In any state, if the query is fetching at any time (including background refetching) isFetching will be true. |

For most queries, it's usually sufficient to check for the isPending state, then the isError state, then finally, assume that the data is available and render the successful state:


```vue
<template>
  <span v-if="isPending">Loading...</span>
  <span v-else-if="isError">Error: {{ error.message }}</span>
  <!-- We can assume by this point that `isSuccess === true` -->
  <ul v-else-if="data">
    <li v-for="todo in data" :key="todo.id">{{ todo.title }}</li>
  </ul>
</template>
```
Status equivalent: 

```vue
<template>
  <span v-if="status === 'pending'">Loading...</span>
  <span v-else-if="status === 'error'">Error: {{ error.message }}</span>
  <!-- also status === 'success', but "else" logic works, too -->
  <ul v-else-if="data">
    <li v-for="todo in data" :key="todo.id">{{ todo.title }}</li>
  </ul>
</template>
```

This should be seen just as an example.

What I really want to see is, that *instead* of removing those neat little computed flags:

> It would be great to consider removing pending for consistency, which currently has a duplicated source of truth.

We should make them usable in both ways, interchangeable without any difference - e.g. make ONE single source of thruth and expose computeds like explained above from them.

**Comment by tlloydthwaites** [comment]
I'm a bit confused about why this is considered an improvement. Is `pending` flag a separate source of truth, or another way of accessing the same source of truth? If `pending` is always true when `status == 'pending'`, I don't see that removing this flag adds any benefit, but does add some rather tedious plumbing for those of us that use it.

**Comment by kyng-cytro** [comment]
Has removal been confirmed? I just got deprecation warnings all over my codebase after an update.

**Comment by tlloydthwaites** [comment]
> Has removal been confirmed? I just got deprecation warnings all over my codebase after an update.

Just deprecated for now, that's how I found out about it. For a breaking change I question whether it really supplies new/better functionality or prevents any problems.

The way I see it, pending means "I am doing something", but status means "here is what I'm doing". Like at a restaurant - I don't really care _what_ you're doing, I just want to know when my meal is ready! :) Honestly I've never used the status, as all I need is to feed a loading spinner (and I'd wager this is true for many). Having to manually build a computed prop every time seems a bit awkward to me, and I'm not seeing any problem this fixes.

**Comment by manniL** [comment]
Hey all 👋 

We highly appreciate your feedback on the topic 🙏 

Eventually, we decided to **not remove** `pending` in Nuxt 4 but to update the implementation to match `status === 'pending'`. This will still be a breaking change and people using `useAsyncData` or `useFetch` with `immediate: false` might have to update their implementation.

We are planning to remove `pending` in Nuxt 5 though.

**Comment by MickL** [comment]
I am a bit confused. I replaced `pending` with `status === 'pending'` in all my components. Unfortunately the app stopped working.

The right choice might be replacing `pending` with `status.value === "idle" || status.value === "pending" || !data.value` but is this really desired? Isnt it too cumberstone?

**Comment by manniL** [comment]
> I am a bit confused. I replaced `pending` with `status === 'pending'` in all my components. Unfortunately the app stopped working.

What do you mean with "stopped working"?

> The right choice might be replacing `pending` with `status.value === "idle" || status.value === "pending" || !data.value` but is this really desired? Isnt it too cumberstone?

This is the current behavior which isn't necessarily correct for a "pending" state.

**Comment by tlloydthwaites** [comment]
I still haven't seen a clear justification for this. Nobody has yet convincingly explained how this would improve anything, or what potential pitfalls it would fix. However, I think it is becoming evident that it would require extensive migration efforts for existing codebases, and then require pointlessly repetitive boilerplate to get the same functionality - all for nil actual change in performance, security, or maintainability. (Remember that employees have to justify these hours to their bosses - what business case could they make for spending that time/money?)

I would say that if you have a semantic or stylistic problem with the pending flag, then... _don't use it_. Wrap `useFetch` in your own composable and make it report its status however you want. However, considering that Nuxt is a large, well-regarded and highly-deployed project that people rely on, I think tinkering with an established API, especially one so extensively used, should be approached with extreme caution and extensive justification. Changes like these have real ramifications for businesses, and we should weigh up carefully whether it is essential or a matter of taste.

**Comment by tlloydthwaites** [comment]
> TL;DR - only when using `immediate: false`, `pending` does not behave correctly.

True, I did get a surprise when I first used `immediate: false`! I figure that pending means "there's no data yet", but if it now means "I am actually talking to the server right now" that makes sense.

**Comment by manniL** [comment]
> True, I did get a surprise when I first used `immediate: false`! I figure that pending means "there's no data yet", but if it now means "I am actually talking to the server right now" that makes sense.

Exactly! `pending` should mean "the request started but didn't finish yet" and not only "it didn't finish yet" 👍

**Comment by mrleblanc101** [comment]
Replace a propriety that return a Boolean with a propriety that return a Dictionnary of Strings is not ideal for people who are not using TypeScript.

**Comment by Anoesj** [comment]
While I agree that the old implementation of `pending` is wrong after reading this discussion, updating the implementation while also deprecating `pending` is a weird decision. This solution would have been better IMHO:
- Deprecate `pending` and keep the faulty implementation.
- Mention the caveats in the docs. Right now, [the docs](https://nuxt.com/docs/api/composables/use-async-data) don't even mention `pending`.
- To make it easier to find out why `pending` is deprecated, add a link to an explanation (in the docs) in the JSDoc `@description` of `pending`.

By the way, I did not read about this breaking change in the [breaking changes](https://nuxt.com/docs/getting-started/upgrade) of Nuxt 4?

**Comment by Hebilicious** [comment]
@manniL
It looks like there's 2 groups of users for pending : 

1) Those who want a computed flag that is equivalent to `status === 'pending'`
2) Those who want a shortcut for `status.value === "idle" || status.value === "pending" || !data.value`

Maybe a way to satisfy both set of users would be keep the semantically incorrect name `pending`and to deprecate it as @Anoesj.

We could introduce computed flags such as `isPending` that would return `status === 'pending'` for the 1st set of users but I personally think `status === 'pending'` is more explicit.

And if you really want you can use custom helpers or computed :

 ```ts
const { status, data } = useFetch(stuff)
const p1 = oldPending({ status, data })
const p2 = pending(status) 
```

But I think the best solution will appear when we understand what exactly the people from group 2 are doing with this logic, and why it can't be replaced by a simple `status === 'pending'` in their cases.
Maybe there's a better way to serve their use case than to introduce `isPendingOrIdleOrEmpty` ?
Are they actyally trying to display the exact same content in 'pending' and 'idle' and 'no data' ?
I would assume that in most cases you'd want some variation of this :

```vue
<template>
  <span v-if="status === 'pending'">Loading...</span>
  <span v-else-if="status === 'success'">Success </span>
  <span v-else-if="status === 'error'">Error </span>
  <span v-else>Fallback</span>
</template>
```
